language/predefined_spec.rb

```diff
@@ -621,6 +621,30 @@ def foo
     end
   end
 
+  ruby_version_is ""..."3.4" do
+    it "can only be assigned an array of String" do
+      begin
+        raise
+      rescue
+        -> {
+          $@ = 123
+        }.should raise_error(TypeError, "backtrace must be Array of String")
+      end
+    end
+  end
+
+  ruby_version_is "3.4" do
+    it "can only be assigned an array of String or an Array of Thread::Backtrace::Location" do
+      begin
+        raise
+      rescue
+        -> {
+          $@ = 123
+        }.should raise_error(TypeError, "backtrace must be an Array of String or an Array of Thread::Backtrace::Location")
+      end
+    end
+  end
+
```

Comment:
The title seems a bit misleading. The tests don't check what can be assigned, but what exception is raised.

Reply:
Also I would consider having only one test case and moving the version guards inside it, e.g.

```ruby
it "..." do
  ruby_version_is ""..."3.4" do
    message = "..."
  end
  ruby_version_is "3.4" do
    message = "..."
  end

  -> {  }.should raise_error(TypeError, message)
end
```

because it seems to me the exception message is kind of an implementation detail here.

Reply:
I have completely refactored this spec: extract the specs for `Exception#set_location` into a shared spec, and use that shared spec to test assignment to `$@`.